File: modules/crash-reporting/android/src/main/java/expo/modules/crashreporting/CrashReportingModule.kt
package expo.modules.crashreporting

import android.content.Context
import android.content.pm.PackageManager
import android.util.AtomicFile
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import io.sentry.Breadcrumb
import io.sentry.IConnectionStatusProvider.ConnectionStatus
import io.sentry.Sentry
import io.sentry.android.core.SentryAndroid
import java.io.File

class CrashReportingModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("CrashReporting")
    // Consent lookup, cache cleanup, and SDK startup do disk work; keep them off the JS thread.
    AsyncFunction("configure") { dsn: String, isProduction: Boolean, version: String ->
      CrashReportingState.configure(
        requireNotNull(appContext.reactContext).applicationContext, dsn, isProduction, version
      )
    }
    Function("getStatus") { CrashReportingState.status() }
    AsyncFunction("setConsent") { enabled: Boolean -> CrashReportingState.setConsent(enabled) }
  }
}

internal object CrashReportingState {
  private var configured = false
  @Volatile private var consentVersion = ""
  private var dsn = ""
  private var canCapture = false
  private var captureStartedAt = 0L
  @Volatile private var granted = false
  @Volatile private var active = false
  @Volatile private var initialization = "not_started"
  @Volatile var breadcrumbCodes: Set<String> = emptySet()
    private set
  @Volatile var breadcrumbLimit = 20
    private set

  private lateinit var context: Context
  private val consentFile get() = File(context.noBackupFilesDir, "cherry-crash-reporting-consent")
  private val cacheDir get() = File(context.cacheDir, "cherry-crash-reporting")

  @Synchronized
  fun configure(context: Context, dsn: String, isProduction: Boolean, version: String): Map<String, Any> {
    return try {
      configureOwner(context, dsn, isProduction, version)
    } catch (error: Throwable) {
      recordStartupFailure()
      configured = false
      throw error
    }
  }

  fun recordStartupFailure() {
    active = false
    granted = false
    initialization = "failed"
    Sentry.close()
  }

  private fun configureOwner(context: Context, dsn: String, isProduction: Boolean, version: String): Map<String, Any> {
    this.context = context
    val metadata = context.packageManager.getApplicationInfo(
      context.packageName, PackageManager.GET_META_DATA
    ).metaData
    breadcrumbCodes = (metadata?.getString("CherryCrashReportingBreadcrumbs") ?: "")
      .split('|').filter { it.isNotEmpty() }.toSet()
    breadcrumbLimit = metadata?.getInt("CherryCrashReportingMaxBreadcrumbs", 20) ?: 20
    this.dsn = dsn
    canCapture = isProduction && dsn.isNotEmpty()
    if (configured) {
      if (version != consentVersion || (!canCapture && active)) {
        consentVersion = version
        revoke()
      } else if (canCapture && granted && !active) {
        start()
      }
      return status()
    }
    configured = true
    consentVersion = version
    // Apply the default only on first use. Preserve AtomicFile's backup of an existing choice too.
    if (version.isNotEmpty() && !consentFile.exists() && !File("${consentFile.path}.bak").exists()) {
      cleanCaches()
      captureStartedAt = System.currentTimeMillis()
      persistConsent()
    }
    val saved = runCatching {
      AtomicFile(consentFile).readFully().toString(Charsets.UTF_8).split('\n')
    }.getOrDefault(emptyList())
    val startedAt = saved.getOrNull(1)?.toLongOrNull()
    // Unreadable or obsolete records never silently re-enable reporting.
    if (version.isNotEmpty() && saved.size == 2 && saved[0] == version && startedAt != null && startedAt > 0L) {
      granted = true
      captureStartedAt = startedAt
    } else {
      // Nothing recorded without a grant for this disclosure may be sent, including legacy reports.
      cleanCaches()
    }
    if (granted && canCapture) start()
    initialization = if (active) "ready" else if (granted && canCapture) "failed" else "inactive"
    return status()
  }

  @Synchronized
  fun setConsent(enabled: Boolean): Map<String, Any> {
    if (!enabled) {
      revoke()
      return status()
    }
    if (!granted) {
      // The grant time rejects system ANR history from before consent on later launches.
      captureStartedAt = System.currentTimeMillis()
      persistConsent()
      granted = true
    }
    if (canCapture && !active) start()
    return status()
  }

  private fun persistConsent(enabled: Boolean = true) {
    val file = AtomicFile(consentFile)
    val output = file.startWrite()
    try {
      val value = if (enabled) "$consentVersion\n$captureStartedAt" else "disabled"
      output.write(value.toByteArray(Charsets.UTF_8))
      file.finishWrite(output)
    } catch (error: Throwable) {
      file.failWrite(output)
      throw error
    }
  }

  private fun revoke() {
    // The gates close before persistence, SDK shutdown, and cache cleanup.
    granted = false
    active = false
    initialization = "inactive"
    Sentry.close()
    persistConsent(false)
    cleanCaches()
  }

  fun status(): Map<String, Any> = mapOf(
    "enabled" to granted, "active" to (active && granted),
    "consentVersion" to consentVersion, "initialization" to initialization
  )

  private fun addBreadcrumb(code: String) {
    if (!active || !granted || code !in breadcrumbCodes) return
    Sentry.addBreadcrumb(Breadcrumb().apply {
      category = "app.diagnostic"
      message = code
    })
  }

  private fun cleanCaches() {
    check(!cacheDir.exists() || cacheDir.deleteRecursively()) { "Could not remove crash reporting cache" }
    // Legacy reports were captured before this app recorded consent.
    val legacy = File(context.cacheDir, "sentry")
    check(!legacy.exists() || legacy.deleteRecursively()) { "Could not remove legacy crash reports" }
  }

  private fun start() {
    active = true
    initialization = "starting"
    SentryAndroid.init(context) { options ->
      options.dsn = dsn
      options.environment = "production"
      options.cacheDirPath = cacheDir.absolutePath
      options.isSendDefaultPii = false
      options.isSendClientReports = false
      options.maxBreadcrumbs = breadcrumbLimit
      options.setBeforeBreadcrumb { breadcrumb, _ -> sanitizeCrashBreadcrumb(breadcrumb) }
      options.isEnableAutoSessionTracking = false
      options.isEnableActivityLifecycleBreadcrumbs = false
      options.isEnableAppLifecycleBreadcrumbs = false
      options.isEnableSystemEventBreadcrumbs = false
      options.isEnableAppComponentBreadcrumbs = false
      options.isEnableNetworkEventBreadcrumbs = false
      options.isEnableUserInteractionBreadcrumbs = false
      options.isEnableAutoActivityLifecycleTracing = false
      options.isEnableUserInteractionTracing = false
      options.isEnableFramesTracking = false
      options.isAttachScreenshot = false
      options.isAttachViewHierarchy = false
      options.logs.isEnabled = false
      options.isEnableNdk = true
      options.isEnableScopeSync = false
      options.setTransportGate {
        active && granted &&
          options.connectionStatusProvider.connectionStatus != ConnectionStatus.DISCONNECTED
      }
      options.setBeforeSend { event, _ ->
        // Android can recover ANRs from OS history even when Sentry was not running at the time.
        if (active && granted && event.timestamp.time >= captureStartedAt) {
          sanitizeCrashEvent(event)
        } else null
      }
    }
    if (!Sentry.isEnabled()) active = false
    initialization = if (active) "ready" else "failed"
    addBreadcrumb("startup.native")
  }
}

File: modules/crash-reporting/ios/CrashReportingModule.swift
import ExpoModulesCore
import Foundation
import Sentry

// An OTA or JS reload may close this gate, but can never enable a non-production binary.
private let isCrashReportingAllowed: Bool = {
  #if DEBUG
  return false
  #else
  return Bundle.main.object(forInfoDictionaryKey: "CherryCrashReportingEnabled") as? Bool == true
  #endif
}()

public class CrashReportingModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CrashReporting")
    // Consent lookup, cache cleanup, and SDK startup do disk work; keep them off the JS thread.
    AsyncFunction("configure") {
      (dsn: String, isProduction: Bool, version: String) throws -> [String: Any] in
      try CrashReportingState.shared.configure(
        dsn: dsn, isProduction: isProduction, version: version)
    }
    Function("getStatus") { CrashReportingState.shared.status() }
    AsyncFunction("setConsent") { (enabled: Bool) throws -> [String: Any] in
      try CrashReportingState.shared.setConsent(enabled)
    }
  }
}

public class CrashReportingAppDelegateSubscriber: ExpoAppDelegateSubscriber {
  // willFinish runs before the app's didFinish callback creates the React Native root.
  public func application(
    _ application: UIApplication,
    willFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]? = nil
  ) -> Bool {
    let config = Bundle.main.infoDictionary ?? [:]
    // Configuration records a fixed failure state; reporting must never prevent app startup.
    _ = try? CrashReportingState.shared.configure(
      dsn: config["CherryCrashReportingDsn"] as? String ?? "",
      isProduction: isCrashReportingAllowed,
      version: config["CherryCrashReportingConsentVersion"] as? String ?? ""
    )
    return true
  }
}

private final class CrashReportingState {
  static let shared = CrashReportingState()
  private let controlLock = NSLock()
  private let gateLock = NSLock()
  private var configured = false
  private var consentVersion = ""
  private var dsn = ""
  private var canCapture = false
  // Read by SDK callbacks under gateLock; written only while holding controlLock.
  private var granted = false
  private var active = false
  private var initialization = "not_started"

  func configure(dsn: String, isProduction: Bool, version: String) throws -> [String: Any] {
    controlLock.lock()
    var succeeded = false
    defer {
      if !succeeded {
        setGate(granted: false, active: false)
        SentrySDK.close()
        gateLock.withLock { initialization = "failed" }
        configured = false
      }
      controlLock.unlock()
    }
    self.dsn = dsn
    canCapture = isCrashReportingAllowed && isProduction && !dsn.isEmpty
    if configured {
      if version != consentVersion || (!canCapture && isActive) {
        gateLock.withLock { consentVersion = version }
        try revoke()
      } else if canCapture && isGranted && !isActive {
        start()
      }
      succeeded = true
      return status()
    }
    configured = true
    gateLock.withLock { consentVersion = version }
    let file = try consentFile()
    // Apply the default only on first use. Never overwrite an opt-out or an existing policy record.
    if !version.isEmpty && !FileManager.default.fileExists(atPath: file.path) {
      try cleanCaches()
      try version.write(to: file, atomically: true, encoding: .utf8)
    }
    if !version.isEmpty && (try? String(contentsOf: file, encoding: .utf8)) == version {
      setGate(granted: true, active: false)
    } else {
      // Nothing recorded without a grant for this disclosure may be sent, including legacy reports.
      try cleanCaches()
    }
    if isGranted && canCapture {
      start()
    }
    gateLock.withLock {
      initialization = active ? "ready" : (granted && canCapture ? "failed" : "inactive")
    }
    succeeded = true
    return status()
  }

  func setConsent(_ enabled: Bool) throws -> [String: Any] {
    controlLock.lock()
    defer { controlLock.unlock() }
    if !enabled {
      try revoke()
      return status()
    }
    if !isGranted {
      try consentVersion.write(to: consentFile(), atomically: true, encoding: .utf8)
      setGate(granted: true, active: isActive)
    }
    if canCapture && !isActive {
      start()
    }
    return status()
  }

  private func revoke() throws {
    // The gates close before SDK shutdown and disk cleanup.
    setGate(granted: false, active: false)
    SentrySDK.close()
    try "disabled".write(to: consentFile(), atomically: true, encoding: .utf8)
    try cleanCaches()
  }

  func status() -> [String: Any] {
    gateLock.withLock {
      [
        "enabled": granted, "active": active && granted,
        "consentVersion": consentVersion, "initialization": initialization,
      ]
    }
  }

  private func addBreadcrumb(_ code: String) {
    guard isActive && isGranted, crashBreadcrumbCodes.contains(code) else { return }
    let crumb = Breadcrumb(level: .info, category: "app.diagnostic")
    crumb.message = code
    SentrySDK.addBreadcrumb(crumb)
  }

  private var isGranted: Bool { gateLock.withLock { granted } }
  private var isActive: Bool { gateLock.withLock { active } }

  private func setGate(granted: Bool, active: Bool) {
    gateLock.withLock {
      self.granted = granted
      self.active = active
      initialization = active ? "starting" : "inactive"
    }
  }

  private func consentFile() throws -> URL {
    var directory = try FileManager.default.url(
      for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true
    )
    .appendingPathComponent("cherry-crash-reporting", isDirectory: true)
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    var values = URLResourceValues()
    values.isExcludedFromBackup = true
    try directory.setResourceValues(values)
    return directory.appendingPathComponent("consent")
  }

  private var caches: URL {
    FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
  }

  private var cacheDirectory: URL {
    caches.appendingPathComponent("cherry-crash-reporting", isDirectory: true)
  }

  private func cleanCaches() throws {
    // io.sentry and SentryCrash hold reports from the old, unconditional initialization.
    let directories = [cacheDirectory] + ["io.sentry", "SentryCrash"].map(caches.appendingPathComponent)
    for directory in directories where FileManager.default.fileExists(atPath: directory.path) {
      try FileManager.default.removeItem(at: directory)
    }
  }

  private func start() {
    setGate(granted: true, active: true)
    SentrySDK.start { options in
      options.dsn = self.dsn
      options.environment = "production"
      options.cacheDirectoryPath = self.cacheDirectory.path
      options.sendDefaultPii = false
      options.experimental.enableLogs = false
      options.sendClientReports = false
      options.maxBreadcrumbs = UInt(crashBreadcrumbLimit)
      options.beforeBreadcrumb = { sanitizeCrashBreadcrumb($0) }
      options.enableAutoBreadcrumbTracking = false
      options.enableNetworkBreadcrumbs = false
      options.enableSwizzling = false
      options.enableAutoPerformanceTracing = false
      options.enableNetworkTracking = false
      options.enableFileIOTracing = false
      options.enableCaptureFailedRequests = false
      options.enableAutoSessionTracking = false
      options.attachScreenshot = false
      options.attachViewHierarchy = false
      options.beforeSend = { [weak self] event in
        guard let self, self.gateLock.withLock({ self.active && self.granted }) else { return nil }
        return sanitizeCrashEvent(event)
      }
    }
    if !SentrySDK.isEnabled {
      setGate(granted: true, active: false)
      gateLock.withLock { initialization = "failed" }
    } else {
      gateLock.withLock { initialization = "ready" }
    }
    addBreadcrumb("startup.native")
  }
}
